### PR TITLE
support sms providers

### PR DESCRIPTION
**Pull request by zhu733756**
Signed-off-by: zhu733756 <talonzhu@yunify.com>

/kind feature

This pr  is a realization of the proposal #76 ,  aims to integrate sms crd to nm, features can be described as followings:
- add SMS-related config definitions  and  config  webhook  with version v2beta2, currently only support aliyun and tencent sms provider.
- added common SMS-related receiver definitions with version v2beta2.
- added common SMS-related struct and methods at `pkg/notify/config/types.go`.
- added a sms notifier package at `pkg/notify/notifier/sms`, and the sms notifier package have provided a factory method can be used for other sms providers.
- added sms aliyun provider, sms tencent provider will be integrated in another commit.
- crds located at `config/*` and `helm/*` will be integrated in later prs.

### CODE REVIEW

**Comment by benjaminhuo** [comment]
@wanjunlei pls help to review this

File: pkg/notify/notifier/sms/aliyun.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	SignName        string
+	NotifierCfg     *config.Config
+	TemplateCode    string
+	AccessKeyId     *v2beta2.Credential
+	AccessKeySecret *v2beta2.Credential
+	PhoneNums       string
+}
+
+func NewAliyunProvider(c *config.Config, providers *v2beta2.Providers, phoneNumbers []string) Provider {
+	phoneNum := strings.Join(phoneNumbers, ",")
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/aliyun.go:24]: Is there any limitation of how many phone numbers are allowed in one API call?

File: pkg/notify/notifier/sms/aliyun.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	config := &openapi.Config{}
+	config.AccessKeyId = &accessKeyId
+	config.AccessKeySecret = &accessKeySecret
+	client, err := dysmsapi.NewClient(config)
+	if err != nil {
+		return fmt.Errorf("cannot make a client with accessKeyId:%s,accessKeySecret:%s", accessKeyId, accessKeySecret)
+	}
+
+	templateParam := `{"code":"` + messages + `"}`
+	sendReq := &dysmsapi.SendSmsRequest{
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/aliyun.go:54]: sendReq => req

File: pkg/notify/notifier/sms/aliyun.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		PhoneNumbers:  &a.PhoneNums,
+		SignName:      &a.SignName,
+		TemplateCode:  &a.TemplateCode,
+		TemplateParam: &templateParam,
+	}
+	sendResp, err := client.SendSms(sendReq)
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/aliyun.go:60]: sendResp => resp

File: pkg/notify/notifier/sms/interface.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	}
+	availableFactoryFuncs[name] = p
+}
+
+func GetProviderFunc(name string) (ProviderFactory, bool) {
+	p, ok := availableFactoryFuncs[name]
+	return p, ok
+}
+
+func GetFirstAvailableProviderFunc() (ProviderFactory, error) {
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/interface.go:37]: There should be GetDefaultProvider ?
GetProviderFunc => GetProvider
GetFirstAvailableProviderFunc => GetFirstAvailableProvider ?

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/apis/v2beta2/config_types.go [hunk 5]
```diff
@@ -120,13 +120,44 @@ type WechatConfig struct {
 	WechatApiSecret *Credential `json:"wechatApiSecret"`
 }
 
+// Sms Aliyun provider parameters
+type AliyunProviderParams struct {
```
Comment (benjaminhuo) [inline pkg/apis/v2beta2/config_types.go:124]: AliyunProviderParams
=>
AliyunSMS

TencentProviderParams
=>
TencentSMS

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/notify/notifier/sms/aliyun.go [hunk 6]
```diff
@@ -58,19 +59,12 @@ func (a *AliyunNotifier) MakeRequest(ctx context.Context, messages string) error
 	}
 	sendResp, err := client.SendSms(sendReq)
 	if err != nil {
-		return err
+		return fmt.Errorf("[Aliyun  SendSms] An API error has returned: %s", err.Error())
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/aliyun.go:62]: An API error has returned
=>
An API error occurs

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/notify/notifier/sms/sms.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		defer func() {
+			_ = level.Debug(n.logger).Log("msg", "SmsNotifier: send message", "used", time.Since(start).String())
+		}()
+
+		newData := utils.FilterAlerts(data, s.Selector, n.logger)
+		if len(newData.Alerts) == 0 {
+			return nil
+		}
+		var value interface{} = newData
+		if n.templateName != DefaultTemplate {
```
Comment (wanjunlei) [inline pkg/notify/notifier/sms/sms.go:96]: It is not appropriate to directly use the JSON string of alert as the default message format, maybe `nm.default.text` is more suitable, or you can define a new template.

File: pkg/notify/notifier/sms/sms.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+		var buf bytes.Buffer
+		if err := utils.JsonEncode(&buf, value); err != nil {
+			_ = level.Error(n.logger).Log("msg", "SmsNotifier: encode message error", "error", err.Error())
+			return err
+		}
+		messages := buf.String()
+
+		// select an available provider function
+		var selectedProviderFunc ProviderFactory
+		if s.SmsConfig.DefaultProvider != "" {
```
Comment (wanjunlei) [inline pkg/notify/notifier/sms/sms.go:115]: These two functions can be combined into one function, the judgment logic is implemented in the function. The readability of the code will be better.
Just a suggestion, will not block this pr.

**Comment by benjaminhuo** [comment]
@zhu733756 pls also add sample yamls to https://github.com/kubesphere/notification-manager/tree/master/config/samples

**Comment by zhu733756** [comment]
All the suggestions have been fixed, pls have a look. @benjaminhuo @wanjunlei

File: pkg/notify/notifier/sms/interface.go [hunk 9]
```diff
@@ -29,16 +29,21 @@ func Register(name string, p ProviderFactory) {
...
+		p, ok := availableFactoryFuncs[name]
+		if !ok {
+			return nil, errors.New("the given default sms provider not registered")
+		}
+		return p, nil
+	} else {
+		// use the first available provider func if the default provider not given
+		for _, p := range availableFactoryFuncs {
+			if p != nil {
+				return p, nil
```
Comment (benjaminhuo) [inline pkg/notify/notifier/sms/interface.go:44]: do not return loop variable directly, assign it to another variable and return that

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: config/samples/sms.yaml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+metadata:
+  labels:
+    app: notification-manager
+    type: global
+  name: global-sms-receiver
+spec:
+  sms:
+    enabled: true
+    phoneNumbers:
+    - "13627275733"
```
Comment (benjaminhuo) [inline config/samples/sms.yaml:13]: do not use a real phone number here

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: config/samples/sms.yaml (hunk 10 above)
Comment (zhu733756) [reply config/samples/sms.yaml:13]: fixed

File: pkg/notify/notifier/sms/interface.go (hunk 9 above)
Comment (zhu733756) [reply pkg/notify/notifier/sms/interface.go:44]: fixed

**Review by benjaminhuo — APPROVED** [review]
/lgtm

File: pkg/notify/notifier/sms/sms.go (hunk 7 above)
Comment (zhu733756) [reply pkg/notify/notifier/sms/sms.go]: fixed, pls have a look

**Comment by zhu733756** [comment]
Currently, the default provider changed, the sms notifier still use the older provider.  Add a new commit to fix it, pls have a look.

**Review by wanjunlei — APPROVED** [review]

**Review by benjaminhuo — APPROVED** [review]

**Comment by zhu733756** [comment]
> Currently, the default provider changed, the sms notifier still use the older provider. Add a new commit to fix it, pls have a look.

This commit would remove because the test scenario occurs error. The old code included this scenario.

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**